File: bending-api/src/main/java/pw/dotdash/bending/api/bender/Bender.java
package pw.dotdash.bending.api.bender;

import org.spongepowered.api.data.DataSerializable;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.util.Identifiable;
import pw.dotdash.bending.api.ability.Ability;
import pw.dotdash.bending.api.ability.AbilityExecutionType;
import pw.dotdash.bending.api.ability.AbilityTask;
import pw.dotdash.bending.api.ability.AbilityType;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * A Bender represents an in-game entity that can execute bending abilities.
 */
public interface Bender extends Identifiable, DataSerializable {

    /**
     * Gets the player associated with the bender.
     *
     * @return The associated player, if available
     */
    Optional<Player> getPlayer();

    /**
     * Gets the ability hotbar that directly maps to the bender's real hotbar.
     *
     * @return The ability hotbar
     */
    List<Optional<Ability>> getHotbar();

    /**
     * Gets the currently selected ability.
     *
     * @return The selected ability, if available
     */
    Optional<Ability> getSelectedAbility();

    /**
     * Sets the currently selected ability.
     *
     * @param ability The ability to set
     * @return True if the operation was successful
     */
    boolean setSelectedAbility(@Nullable Ability ability);

    /**
     * Gets the ability equipped at the given hotbar index.
     *
     * <p>Note: index must be within zero to eight.</p>
     *
     * @param hotbarIndex The index to query
     * @return The equipped ability, if available
     */
    Optional<Ability> getEquippedAbility(int hotbarIndex);

    /**
     * Sets the ability equipped at the given hotbar index.
     *
     * <p>Note: index must be within zero to eight.</p>
     *
     * @param hotbarIndex The index to set at
     * @param ability     The ability to set
     * @return True if the operation was successful
     */
    boolean setEquippedAbility(int hotbarIndex, @Nullable Ability ability);

    /**
     * Removes all equipped abilities.
     */
    void clearEquippedAbilities();

    /**
     * Gets all currently running abilities' {@link AbilityTask}s.
     *
     * @return All currently running abilities
     */
    Collection<AbilityTask> getRunningAbilities();

    /**
     * Executes the ability by the given execution type.
     *
     * @param ability       The ability to execute
     * @param executionType The way to execute the ability
     * @return The ability task, if successful
     */
    Optional<AbilityTask> execute(Ability ability, AbilityExecutionType executionType);

    CompletableFuture<Void> waitForExecution(AbilityType type, AbilityExecutionType executionType);

    boolean cancel(AbilityType type);

    boolean hasCooldown(AbilityType type);

    boolean setCooldown(AbilityType type, long durationMilli);

    long removeCooldown(AbilityType type);
}

File: src/main/kotlin/arvenwood/bending/api/Bender.kt
package arvenwood.bending.api

import arvenwood.bending.api.ability.Ability
import arvenwood.bending.api.ability.AbilityExecutionType
import arvenwood.bending.api.ability.AbilityJob
import arvenwood.bending.api.ability.AbilityType
import kotlinx.coroutines.Job
import org.spongepowered.api.entity.living.player.Player

interface Bender {

    /**
     * The underlying [Player] that this bender represents.
     */
    val player: Player

    /**
     * The 9 currently equipped abilities, or null if that spot has none.
     */
    val equippedAbilities: List<Ability<*>?>

    /**
     * The currently selected ability (as in the player's current slot index in the [equippedAbilities]).
     */
    var selectedAbility: Ability<*>?

    /**
     * Gets all currently running abilities from this bender.
     */
    val runningAbilities: Collection<AbilityJob>

    /**
     * Gets the currently equipped ability at the given hotbar index, if any.
     */
    fun getEquipped(hotbarIndex: Int): Ability<*>?

    /**
     * Sets the equipped ability at the given hotbar index, or null to remove one.
     */
    fun setEquipped(hotbarIndex: Int, ability: Ability<*>?)

    /**
     * Removes all equipped abilities.
     */
    fun clearEquipped()

    /**
     * Executes the given ability by the given execution type.
     */
    fun execute(ability: Ability<*>, executionType: AbilityExecutionType)

    /**
     * Waits for the bender to execute the given ability by the given execution type.
     */
    suspend fun awaitExecution(type: AbilityType<*>, executionType: AbilityExecutionType)

    /**
     * Waits for the bender to execute the given ability by the given execution type.
     */
    fun deferExecution(type: AbilityType<*>, executionType: AbilityExecutionType): Job

    /**
     * Cancels any running/waiting abilities for the given ability type.
     */
    fun cancel(type: AbilityType<*>): Boolean

    /**
     * Checks if this bender is on cooldown for the given ability.
     *
     * @param type The ability to check
     * @return Whether the ability is on cooldown for this bender
     */
    fun hasCooldown(type: AbilityType<*>): Boolean

    /**
     * Sets the duration of which this bender can't use the given ability.
     *
     * @param type The ability to set for
     * @param duration How long they can't use it for
     */
    fun setCooldown(type: AbilityType<*>, duration: Long)

    /**
     * Removes the cooldown for the given ability.
     *
     * @param type The ability to remove for
     * @return When the ability would have been usable again, or null if no cooldown
     */
    fun removeCooldown(type: AbilityType<*>): Long?
}